src/Sarif/Writers/SarifConsolidator.cs

```diff
@@ -0,0 +1,259 @@
+
+            // Remove Result components if configured to
+            result.CodeFlows = (this.RemoveCodeFlows ? null : result.CodeFlows);
+            result.RelatedLocations = (this.RemoveRelatedLocations ? null : result.RelatedLocations);
+            result.Graphs = (this.RemoveGraphs ? null : result.Graphs);
+            result.GraphTraversals = (this.RemoveGraphTraversals ? null : result.GraphTraversals);
+            result.Stacks = (this.RemoveStacks ? null : result.Stacks);
+            result.WebRequest = (this.RemoveWebRequests ? null : result.WebRequest);
+            result.WebResponse = (this.RemoveWebResponses ? null : result.WebResponse);
+
+            // Truncate long Messages if configured
+            if (this.MessageLengthLimitBytes > 0)
+            {
+                if (result?.Message?.Text?.Length > this.MessageLengthLimitBytes)
+                {
+                    result.Message.Text = result.Message.Text.Substring(0, this.MessageLengthLimitBytes.Value);
+                }
+            }
+
+            // Trim Location and RelatedLocation properties
+            result.Locations = Trim(result.Locations);
+            result.RelatedLocations = Trim(result.RelatedLocations);
+
+            // Trim and Consolidate CodeFlows and ThreadFlows
+            ConsolidateCodeFlows(result);
+        }
+
+        public void ConsolidateCodeFlows(Result result)
+        {
+            if (result.CodeFlows != null)
+            {
+                foreach (CodeFlow cFlow in result.CodeFlows)
+                {
+                    foreach (ThreadFlow tFlow in cFlow.ThreadFlows)
+                    {
+                        Consolidate(tFlow);
+                    }
+                }
+            }
+        }
+
+        public void Consolidate(ThreadFlow tFlow)
+        {
+            if (tFlow.Locations != null)
+            {
+                TotalThreadFlowLocations += tFlow.Locations.Count;
+
+                for (int i = 0; i < tFlow.Locations.Count; ++i)
+                {
+                    ThreadFlowLocation tfl = tFlow.Locations[i];
+                    Trim(tfl.Location);
+
+                    if (tfl.Index == -1)
+                    {
+                        if (!_uniqueThreadFlows.TryGetValue(tfl, out int tflIndex))
+                        {
+                            tflIndex = _run.ThreadFlowLocations.Count;
+                            _run.ThreadFlowLocations.Add(tfl);
+                            _uniqueThreadFlows[tfl] = tflIndex;
+                        }
+
+                        tFlow.Locations[i] = new ThreadFlowLocation() { Index = tflIndex };
+                    }
+                }
+            }
+        }
+
+        public IList<Location> Trim(IList<Location> locations)
+        {
+            if (locations == null || locations.Count == 0) { return locations; }
+
+            HashSet<Location> uniqueLocations = new HashSet<Location>(Location.ValueComparer);
```

Comment:
I added de-duplication logic for both location List types. I wanted the output to tell me how many ThreadFlowLocations I consolidated (by moving them to Run.ThreadFlowLocations) and how many other locations I consolidated (by removing duplicates within any single result). I didn't think breaking out Locations and LogicalLocations was worthwhile, because it was the same functional change.

I could separate them, but again, it's just to give the user output giving a sense of how much cleanup occurred.